### PR TITLE
Allow multi-arch tags in MCR tags metadata templates

### PR DESCRIPTION
**Pull request by lbussell**
This change allows patterns like:

`aspire-dashboard-tags.yml`
```yaml
$(McrTagsYmlRepo:aspire-dashboard)
$(McrTagsYmlTagGroup:8)
```

`monitor-tags.yml`
```yaml
$(McrTagsYmlRepo:monitor)
$(McrTagsYmlTagGroup:9.0)
$(McrTagsYmlTagGroup:8.1-preview-ubuntu-chiseled|.NET Monitor Preview Tags)
$(McrTagsYmlTagGroup:8.1-preview-cbl-mariner-distroless|.NET Monitor Preview Tags)
$(McrTagsYmlTagGroup:8.0-ubuntu-chiseled)
$(McrTagsYmlTagGroup:8.0-cbl-mariner-distroless)
$(McrTagsYmlTagGroup:6.3-alpine)
$(McrTagsYmlTagGroup:6.3-ubuntu-chiseled)
$(McrTagsYmlTagGroup:6.3-cbl-mariner)
$(McrTagsYmlTagGroup:6.3-cbl-mariner-distroless)
```

`runtime-tags.yml`:
```yaml
$(McrTagsYmlRepo:runtime)
$(McrTagsYmlTagGroup:9.0-bookworm-slim)
$(McrTagsYmlTagGroup:9.0-alpine3.20)
$(McrTagsYmlTagGroup:9.0-noble)
$(McrTagsYmlTagGroup:9.0-noble-chiseled)
$(McrTagsYmlTagGroup:9.0-noble-chiseled-extra)
$(McrTagsYmlTagGroup:9.0-azurelinux3.0)
$(McrTagsYmlTagGroup:9.0-azurelinux3.0-distroless)
$(McrTagsYmlTagGroup:9.0-azurelinux3.0-distroless-extra)
$(McrTagsYmlTagGroup:8.0-bookworm-slim)
<snip>
```

This works by checking if the tag in the variable is multi-arch or not, and if so, adds yaml tag groups for each architecture. This meant we needed a different solution for custom tag tables - I added the option to put that inline with the variable (which you can see above).

I made sure that using the multi-arch tags didn't change the readme output or sorting. In order to do that I had to add some sorting logic to the readme tags table generation - but it's just automating what we were already doing by hand.

### CODE REVIEW

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs [hunk 1]
```diff
@@ -153,10 +163,19 @@ private static string GenerateTables(string repoName, string tagsMetadata)
         private static string GenerateTables(IEnumerable<TagGroup> tagGroups)
         {
             StringBuilder tables = new();
-            IEnumerable<IGrouping<(string OS, string Architecture, string? OsVersion), TagGroup>> tagGroupsGroupedByOsArch = tagGroups
-                .GroupBy(tagGroup => (tagGroup.OS, tagGroup.Architecture, tagGroup.OS == "windows" ? tagGroup.OsVersion : null));
-            bool isFirstTable = true;
 
+            IEnumerable<IGrouping<(string OS, string Architecture, string? OsVersion), TagGroup>> tagGroupsGroupedByOsArch =
+                tagGroups
+                    // Linux should be listed before Windows, then sort by Windows versions
```
[the remark this answers is not in the document]
Comment (lbussell) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs:169]: This is needed for when there are multi-arch tags that include Windows. Examples are the samples and .NET 6 images.

An alternative to using this logic here is to blanket ignore all Windows images when generating tag metadatas for multi-arch tags, but that seemed a little too .NET-specific to me at the time I implemented this.

[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs:169]: But why do the tags need to be grouped by OS? That's what I don't understand. The YAML file should declare OS-specific tags. Wouldn't that get rid of the problem with needing to order things beyond arch in Image Builder?

[the remark this answers is not in the document]
Comment (lbussell) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs]: I got this figured out. Ordering by arch in the GenerateReadmes command was resetting the manual OS ordering from the templates. By moving the sorting into the multi-arch tag specific code in McrTagsMetadataGenerator, we can order them in the tags metadata and then just don't mess with the ordering when generating tags tables for readmes.

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs [hunk 2]
```diff
@@ -236,6 +236,12 @@ private static string FormatTagGroupRow(TagGroup tagGroup)
             return row;
         }
 
+        private static int GetWindowsVersionSortingKey(string osVersion)
+        {
+            int version = int.Parse(WindowsVersionRegex.Match(osVersion).Value);
+            return version == 1809 ? 2019 : version;
+        }
```
Comment (mthalman) [inline src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateReadmesCommand.cs:243]: Unused

**Review by mthalman — APPROVED** [review]

**Comment by lbussell** [comment]
@mthalman Review new tests please. Thanks.

**Review by mthalman — APPROVED** [review]

**Pull request merged**